src/components/Filters/index.tsx

```diff
@@ -104,40 +105,55 @@ interface FiltersState {
 	searchString: string;
 	filters: JSONSchema6[];
 	views: FiltersView[];
+	schema: JSONSchema6;
 }
 
 class Filters extends React.Component<FiltersProps, FiltersState> {
 	constructor(props: FiltersProps) {
 		super(props);
-		const { filters, views } = this.props;
+		const { filters, schema, views } = this.props;
+
+		const flatSchema = SchemaSieve.flattenSchema(schema);
 
 		this.state = {
 			showModal: false,
 			searchString: '',
-			edit: [this.getCleanEditModel()],
 			editingFilter: null,
 			filters: filters || [],
 			views: views || [],
+			schema: flatSchema,
+			edit: [this.getCleanEditModel(null, flatSchema)],
 		};
+	}
 
-		// Clean exsting rules on load
+	componentDidUpdate(prevProps: FiltersProps) {
+		// If the schema prop updates, also update the internal 'flat' schema
+		if (!isEqual(prevProps.schema, this.props.schema)) {
+			this.setState({
+				schema: SchemaSieve.flattenSchema(this.props.schema),
+			});
+		}
 	}
 
-	getCleanEditModel(field?: string) {
-		const { schema } = this.props;
+	getCleanEditModel(field?: string | null, schema?: JSONSchema6) {
```

Comment:
Why does this and `getOperators` both now take a schema as a parameter? Doesn't seem to be used here at all. Is this for something else?

Reply:
This is purely to make everything easier when setting up the state in `constructor`. As `getCleanEditModel` and `getOperators` both now use `state` internally, they can't be called before `this.state` has been instantiated. I get around this by allowing you to specify the schema.

Reply:
How about doing something like:
```
const edit = [];
this.state= {
	/*...*/
	schema: flatSchema,
	edit,
};
edit.push(this.getCleanEditModel(null));
```
and drop the added schema property? Would that be too :imp::imp::imp: ?

Reply:
yeah that would be fine